crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -12,6 +11,13 @@ use crate::{
 };
 use thiserror::Error;
 
+/// ### Stack based state machine
+/// Important: `T` must be an enum!
```

Comment:
An `Eq` bound would be correct here. How else are you supposed to determine that the state has changed (I know that you're using the discriminant, but still)?

In general, restrictions that appear only in comments is a bad idea. This one specifically is especially odd if you don't know what's going on. Besides, I can totally see myself using a struct for state values.

Reply:
> I can totally see myself using a struct for state values.

That would be an incorrect use of the API and I'd love prevent it statically, but unfortunately there isn't a trait for enums.

> An Eq bound would be correct here. 

Hmm, ok

Reply:
> That would be an incorrect use of the API [...]

Why? (Assuming the implementation didn't rely on discriminant.)